bin/propolis-server/src/lib/vm/ensure.rs

```diff
@@ -0,0 +1,359 @@
+//! phase. If a higher-level operation fails, it can call a failure handler on
+//! its current phase to unwind the whole operation and drive the VM state
+//! machine to the correct resting state.
+
+use std::sync::Arc;
+
+use propolis_api_types::{
+    instance_spec::{v0::InstanceSpecV0, VersionedInstanceSpec},
+    InstanceEnsureResponse, InstanceMigrateInitiateResponse,
+    InstanceProperties, InstanceSpecEnsureRequest, InstanceState,
+};
+use slog::{debug, info};
+
+use crate::{
+    initializer::{
+        build_instance, MachineInitializer, MachineInitializerState,
+    },
+    vm::request_queue::InstanceAutoStart,
+};
+
+use super::{
+    objects::{InputVmObjects, VmObjects},
+    services::VmServices,
+    state_driver::InputQueue,
+    state_publisher::{ExternalStateUpdate, StatePublisher},
+    EnsureOptions, InstanceEnsureResponseTx, VmError,
+};
+
+/// Holds state about an instance ensure request that has not yet produced any
+/// VM objects or driven the VM state machine to the `ActiveVm` state.
+pub(crate) struct VmEnsureNotStarted<'a> {
+    log: &'a slog::Logger,
+    vm: &'a Arc<super::Vm>,
+    ensure_request: &'a InstanceSpecEnsureRequest,
+    ensure_options: &'a EnsureOptions,
+    ensure_response_tx: InstanceEnsureResponseTx,
+    state_publisher: &'a mut StatePublisher,
+}
+
+impl<'a> VmEnsureNotStarted<'a> {
+    pub(super) fn new(
+        log: &'a slog::Logger,
+        vm: &'a Arc<super::Vm>,
+        ensure_request: &'a InstanceSpecEnsureRequest,
+        ensure_options: &'a EnsureOptions,
+        ensure_response_tx: InstanceEnsureResponseTx,
+        state_publisher: &'a mut StatePublisher,
+    ) -> Self {
+        Self {
+            log,
+            vm,
+            ensure_request,
+            ensure_options,
+            ensure_response_tx,
+            state_publisher,
+        }
+    }
+
+    pub(crate) fn instance_spec(&self) -> &InstanceSpecV0 {
+        let VersionedInstanceSpec::V0(v0) = &self.ensure_request.instance_spec;
+        v0
+    }
+
+    pub(crate) fn state_publisher(&mut self) -> &mut StatePublisher {
+        self.state_publisher
+    }
+
+    /// Creates a set of VM objects using the instance spec stored in this
+    /// ensure request, but does not install them as an active VM.
+    pub(crate) async fn create_objects(
+        self,
+    ) -> anyhow::Result<VmEnsureObjectsCreated<'a>> {
+        debug!(self.log, "creating VM objects");
+
+        let input_queue = Arc::new(InputQueue::new(
+            self.log.new(slog::o!("component" => "request_queue")),
+            match &self.ensure_request.migrate {
+                Some(_) => InstanceAutoStart::Yes,
+                None => InstanceAutoStart::No,
+            },
+        ));
+
+        match initialize_vm_objects_from_spec(
+            self.log,
+            &input_queue,
+            &self.ensure_request.properties,
+            &self.ensure_request.instance_spec,
+            self.ensure_options,
+        )
```

Comment:
not a big deal, but: any reason this can't be a method on `VmEnsureNotStarted`, since it seems like most of its parameters are just fields on `self` that we could avoid passing explicitly?